src/EditorFeatures/CSharp/CommentSelection/CSharpToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,75 @@
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CommentSelection;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Editor.Implementation.CommentSelection;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Operations;
+using Microsoft.VisualStudio.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.CommentSelection
+{
+    /* TODO - Modify these once the toggle block comment handler is added.
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [Name(PredefinedCommandHandlerNames.CommentSelection)]*/
+    internal class CSharpToggleBlockCommentCommandHandler :
+        ToggleBlockCommentCommandHandler
+    {
+        [ImportingConstructor]
+        internal CSharpToggleBlockCommentCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService)
+            : base(undoHistoryRegistry, editorOperationsFactoryService)
+        {
+        }
+
+        protected override async Task<IToggleBlockCommentDocumentDataProvider> GetBlockCommentDocumentData(Document document, ITextSnapshot snapshot,
+            CommentSelectionInfo commentInfo, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            return new CSharpToggleBlockCommentDocumentDataProvider(root);
+        }
+
+        private class CSharpToggleBlockCommentDocumentDataProvider : IToggleBlockCommentDocumentDataProvider
+        {
+            private readonly SyntaxNode _root;
+
+            public CSharpToggleBlockCommentDocumentDataProvider(SyntaxNode root)
+            {
+                _root = root;
+            }
+
+            /// <summary>
+            /// Get a location of itself or the end of the token it is located in.
+            /// </summary>
+            public int GetEmptyCommentStartLocation(int location)
+            {
+                var token = _root.FindToken(location);
+                if (token.Span.Contains(location))
+                {
+                    return token.Span.End;
+                }
+                return location;
+            }
+
+            /// <summary>
+            /// Get the location of the comments from the syntax tree.
+            /// </summary>
+            /// <returns></returns>
+            public IEnumerable<TextSpan> GetBlockCommentsInDocument()
+            {
+                return _root.DescendantTrivia()
+                    .Where(trivia => trivia.IsKind(SyntaxKind.MultiLineCommentTrivia) || trivia.IsKind(SyntaxKind.MultiLineDocumentationCommentTrivia))
+                    .Select(blockCommentTrivia => blockCommentTrivia.Span);
```

Comment:
1. whiile this is hte most correct, i hope we're able to still work if you don't have this.
2. Having this be non-async could be tricky for languages like TS/JS which need to asynchronously call out to an OOP process.  This should also be cancellable.

Reply:
1.  It will work w/out.
2.  Currently it isn't exposed for others outside of roslyn to consume, it's just a csharp addition per the recommendation to avoid IVT

Reply:
i can live with that.  though my preference would be to not even have this code :)  but i'm not going to ask you to remove it.